File: figures/06-search-trees/snippet-red-black-tree.py
class _Node:
    def __init__(self, key, color):
        self.key = key
        self.color = color
        self.left = self.right = self.parent = None


class RedBlackTree:
    def __init__(self):
        self.nil = _Node(0, BLACK)
        self.nil.left = self.nil.right = self.nil.parent = self.nil
        self.root = self.nil
        self._size = 0

    def __len__(self):
        return self._size

    def _find(self, key):
        x = self.root
        while x is not self.nil and x.key != key:
            x = x.left if key < x.key else x.right
        return x

    def contains(self, key):
        return self._find(key) is not self.nil

    def insert(self, key):
        y, x = self.nil, self.root
        while x is not self.nil:
            y = x
            if key == x.key:
                return False
            x = x.left if key < x.key else x.right
        z = _Node(key, RED)
        z.left = z.right = self.nil
        z.parent = y
        if y is self.nil:
            self.root = z
        elif key < y.key:
            y.left = z
        else:
            y.right = z
        self._size += 1
        self._insert_fixup(z)
        return True

    def _rotate_left(self, x):
        y = x.right
        x.right = y.left
        if y.left is not self.nil:
            y.left.parent = x
        y.parent = x.parent
        if x.parent is self.nil:
            self.root = y
        elif x is x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        y.left = x
        x.parent = y

    def _rotate_right(self, y):
        x = y.left
        y.left = x.right
        if x.right is not self.nil:
            x.right.parent = y
        x.parent = y.parent
        if y.parent is self.nil:
            self.root = x
        elif y is y.parent.left:
            y.parent.left = x
        else:
            y.parent.right = x
        x.right = y
        y.parent = x

    def _insert_fixup(self, z):
        while z.parent.color == RED:
            if z.parent is z.parent.parent.left:
                u = z.parent.parent.right
                if u.color == RED:
                    z.parent.color = BLACK
                    u.color = BLACK
                    z.parent.parent.color = RED
                    z = z.parent.parent
                else:
                    if z is z.parent.right:
                        z = z.parent
                        self._rotate_left(z)
                    z.parent.color = BLACK
                    z.parent.parent.color = RED
                    self._rotate_right(z.parent.parent)
            else:
                u = z.parent.parent.left
                if u.color == RED:
                    z.parent.color = BLACK
                    u.color = BLACK
                    z.parent.parent.color = RED
                    z = z.parent.parent
                else:
                    if z is z.parent.left:
                        z = z.parent
                        self._rotate_right(z)
                    z.parent.color = BLACK
                    z.parent.parent.color = RED
                    self._rotate_left(z.parent.parent)
        self.root.color = BLACK
        self.root.parent = self.nil

    def _minimum(self, x):
        while x.left is not self.nil:
            x = x.left
        return x

    def _transplant(self, u, v):
        if u.parent is self.nil:
            self.root = v
        elif u is u.parent.left:
            u.parent.left = v
        else:
            u.parent.right = v
        v.parent = u.parent

    def delete(self, key):
        z = self._find(key)
        if z is self.nil:
            return False
        y = z
        original = y.color
        if z.left is self.nil:
            x = z.right
            self._transplant(z, z.right)
        elif z.right is self.nil:
            x = z.left
            self._transplant(z, z.left)
        else:
            y = self._minimum(z.right)
            original = y.color
            x = y.right
            if y.parent is z:
                x.parent = y
            else:
                self._transplant(y, y.right)
                y.right = z.right
                y.right.parent = y
            self._transplant(z, y)
            y.left = z.left
            y.left.parent = y
            y.color = z.color
        self._size -= 1
        if original == BLACK:
            self._delete_fixup(x)
        if self.root is not self.nil:
            self.root.parent = self.nil
        else:
            self.nil.parent = self.nil
        return True

    def _delete_fixup(self, x):
        while x is not self.root and x.color == BLACK:
            if x is x.parent.left:
                w = x.parent.right
                if w.color == RED:
                    w.color = BLACK
                    x.parent.color = RED
                    self._rotate_left(x.parent)
                    w = x.parent.right
                if w.left.color == BLACK and w.right.color == BLACK:
                    w.color = RED
                    x = x.parent
                else:
                    if w.right.color == BLACK:
                        w.left.color = BLACK
                        w.color = RED
                        self._rotate_right(w)
                        w = x.parent.right
                    w.color = x.parent.color
                    x.parent.color = BLACK
                    w.right.color = BLACK
                    self._rotate_left(x.parent)
                    x = self.root
            else:
                w = x.parent.left
                if w.color == RED:
                    w.color = BLACK
                    x.parent.color = RED
                    self._rotate_right(x.parent)
                    w = x.parent.left
                if w.right.color == BLACK and w.left.color == BLACK:
                    w.color = RED
                    x = x.parent
                else:
                    if w.left.color == BLACK:
                        w.right.color = BLACK
                        w.color = RED
                        self._rotate_left(w)
                        w = x.parent.left
                    w.color = x.parent.color
                    x.parent.color = BLACK
                    w.left.color = BLACK
                    self._rotate_right(x.parent)
                    x = self.root
        x.color = BLACK

    def inorder(self):
        out = []

        def walk(x):
            if x is not self.nil:
                walk(x.left)
                out.append(x.key)
                walk(x.right)

        walk(self.root)
        return out

File: figures/06-search-trees/snippet-red-black-tree.java
    private static final boolean RED=true, BLACK=false;
    private final Node nil=new Node(0,BLACK);
    private Node root=nil;
    private int size;

    private final class Node {
        int key; boolean color; Node left,right,parent;
        Node(int key,boolean color){this.key=key;this.color=color;}
    }
    public RedBlackTree(){nil.left=nil.right=nil.parent=nil;}
    private Node find(int key){Node x=root;while(x!=nil&&x.key!=key)x=key<x.key?x.left:x.right;return x;}
    public boolean contains(int key){return find(key)!=nil;}
    public int size(){return size;}

    public boolean insert(int key){
        Node y=nil,x=root;
        while(x!=nil){y=x;if(key==x.key)return false;x=key<x.key?x.left:x.right;}
        Node z=new Node(key,RED);z.left=z.right=nil;z.parent=y;
        if(y==nil)root=z;else if(key<y.key)y.left=z;else y.right=z;
        size++;insertFixup(z);return true;
    }
    private void insertFixup(Node z){
        while(z.parent.color==RED){
            if(z.parent==z.parent.parent.left){
                Node u=z.parent.parent.right;
                if(u.color==RED){z.parent.color=BLACK;u.color=BLACK;z.parent.parent.color=RED;z=z.parent.parent;}
                else{if(z==z.parent.right){z=z.parent;rotateLeft(z);}z.parent.color=BLACK;z.parent.parent.color=RED;rotateRight(z.parent.parent);}
            }else{
                Node u=z.parent.parent.left;
                if(u.color==RED){z.parent.color=BLACK;u.color=BLACK;z.parent.parent.color=RED;z=z.parent.parent;}
                else{if(z==z.parent.left){z=z.parent;rotateRight(z);}z.parent.color=BLACK;z.parent.parent.color=RED;rotateLeft(z.parent.parent);}
            }
        }
        root.color=BLACK;root.parent=nil;
    }
    private void rotateLeft(Node x){
        Node y=x.right;x.right=y.left;if(y.left!=nil)y.left.parent=x;
        y.parent=x.parent;if(x.parent==nil)root=y;else if(x==x.parent.left)x.parent.left=y;else x.parent.right=y;
        y.left=x;x.parent=y;
    }
    private void rotateRight(Node y){
        Node x=y.left;y.left=x.right;if(x.right!=nil)x.right.parent=y;
        x.parent=y.parent;if(y.parent==nil)root=x;else if(y==y.parent.left)y.parent.left=x;else y.parent.right=x;
        x.right=y;y.parent=x;
    }
    private Node minimum(Node x){while(x.left!=nil)x=x.left;return x;}
    private void transplant(Node u,Node v){
        if(u.parent==nil)root=v;else if(u==u.parent.left)u.parent.left=v;else u.parent.right=v;
        v.parent=u.parent;
    }
    public boolean delete(int key){
        Node z=find(key);if(z==nil)return false;
        Node y=z,x;boolean original=y.color;
        if(z.left==nil){x=z.right;transplant(z,z.right);}
        else if(z.right==nil){x=z.left;transplant(z,z.left);}
        else{
            y=minimum(z.right);original=y.color;x=y.right;
            if(y.parent==z)x.parent=y;
            else{transplant(y,y.right);y.right=z.right;y.right.parent=y;}
            transplant(z,y);y.left=z.left;y.left.parent=y;y.color=z.color;
        }
        size--;if(original==BLACK)deleteFixup(x);
        if(root!=nil)root.parent=nil;else nil.parent=nil;
        return true;
    }
    private void deleteFixup(Node x){
        while(x!=root&&x.color==BLACK){
            if(x==x.parent.left){
                Node w=x.parent.right;
                if(w.color==RED){w.color=BLACK;x.parent.color=RED;rotateLeft(x.parent);w=x.parent.right;}
                if(w.left.color==BLACK&&w.right.color==BLACK){w.color=RED;x=x.parent;}
                else{
                    if(w.right.color==BLACK){w.left.color=BLACK;w.color=RED;rotateRight(w);w=x.parent.right;}
                    w.color=x.parent.color;x.parent.color=BLACK;w.right.color=BLACK;rotateLeft(x.parent);x=root;
                }
            }else{
                Node w=x.parent.left;
                if(w.color==RED){w.color=BLACK;x.parent.color=RED;rotateRight(x.parent);w=x.parent.left;}
                if(w.right.color==BLACK&&w.left.color==BLACK){w.color=RED;x=x.parent;}
                else{
                    if(w.left.color==BLACK){w.right.color=BLACK;w.color=RED;rotateLeft(w);w=x.parent.left;}
                    w.color=x.parent.color;x.parent.color=BLACK;w.left.color=BLACK;rotateRight(x.parent);x=root;
                }
            }
        }
        x.color=BLACK;
    }
    public List<Integer> inorder(){List<Integer>o=new ArrayList<>();inorder(root,o);return List.copyOf(o);}
    private void inorder(Node x,List<Integer>o){if(x!=nil){inorder(x.left,o);o.add(x.key);inorder(x.right,o);}}
